runtime/lua/vim/lsp/util.lua

```diff
@@ -781,34 +781,44 @@ function M.convert_input_to_markdown_lines(input, contents)
   return contents
 end
 
+--- Returns the line/column-based position in `contents` at the given offset.
+---
+---@param offset integer
+---@param contents string[]
+---@return Range2
+local function get_pos_from_offset(offset, contents)
+  local i = 0
+  for l, line in ipairs(contents) do
+    if offset >= i and offset < i + #line then
+      return { l - 1, offset - i + 1 }
+    else
+      i = i + #line + 1
+    end
+  end
+end
+
 --- Converts `textDocument/signatureHelp` response to markdown lines.
 ---
 ---@param signature_help lsp.SignatureHelp Response of `textDocument/SignatureHelp`
 ---@param ft string|nil filetype that will be use as the `lang` for the label markdown code block
 ---@param triggers table|nil list of trigger characters from the lsp server. used to better determine parameter offsets
 ---@return string[]|nil table list of lines of converted markdown.
----@return number[]|nil table of active hl
+---@return Range4|nil table of active hl
```

Comment:
is this correct? `active_hl` and `get_pos_from_offset` appear to have two numbers (actually not a Range2), not Range4.

Reply:
It is correct. Notice how `active_hl` is extended twice with 2 `Range2`s here:
https://github.com/neovim/neovim/blob/6e01e0f7b60f5afe802c6feb52b7f7d5fb15cde7/runtime/lua/vim/lsp/util.lua#L908-L912

Reply:
I see, now I get how it contains four elements. It looked like `active_hl` is `Range4[]` at a glance. Though I believe making it explicit (e.g. `row_start, col_start`, `row_end, `col_end`) would make the code a bit easier to read.

Reply:
I agree that it would make the code easier to read when not inside an editor, but what you described is exactly the definition of `Range4`:
```lua
---@class Range4
---@inlinedoc
---@field [1] integer start row
---@field [2] integer start column
---@field [3] integer end row
---@field [4] integer end column
```

I prefer to use the existing type here, and either way we do want to consolidate the many position/range abstractions we have (see https://github.com/neovim/neovim/issues/25509).

Reply:
Yes, I agree. Just wanted to be clear, I meant using local variables while constructing a table (rather than using list_extend which is not type friendly).

Reply:
I don't get any type errors with the current approach...

Reply:
@wookayin Yep! I left a short comment here:
https://github.com/neovim/neovim/blob/21911f2cc556e6b05ad58a92cc7d4539e6a939ae/runtime/lua/vim/lsp/util.lua#L900-L906

But I'll make sure to clarify this in the test too.